tools/node-modules

```diff
@@ -56,35 +56,96 @@ EOF
     clone_from_cache "${sha}"
 }
 
-cmd_rebuild() {
-    for file in npm-shrinkwrap.json package-lock.json yarn.lock; do
-        if [ -e "${file}" ]; then
-            echo "Please delete ${file} from $(pwd) and start again"
-            exit 1
-        fi
-    done
+cmd_npm_install() {
+    if [ -f /run/.toolboxenv ]; then
+        exec flatpak-spawn --host -- "$0" npm_install "$@"
+        exit 1
+    fi
+
+    CACHE="cockpit-project-npm-cache-volume"
+    IMAGE="docker.io/library/node:alpine"
+
+    # make sure the node user can write to the cache volume
+    podman run \
+        --rm \
+        --volume "${CACHE}":/home/node/.npm:U \
+        "${IMAGE}" chown -R node:node /home/node >&2
+
+    # do the actual work
+    podman run \
```

Comment:
I suggest `-it` here, so that you see proper output and can stop the container with Control-C. It's otherwise a bit annoying.

Reply:
This can't work as suggested.  `--interactive` is `-i` and we're already doing that, but `-t` has some side effects that make it unsuitable for piping binary data, and (particularly) for having separate stdout (`tar`) and stderr (logs) streams.

Reply:
fwiw, I agree that it's a bit annoying.  I consider wrapping the entire thing in `timeout` to make it killable.  I'll put a bit more research into our possibilities here.

Reply:
After a gigantic amount of research I got to the bottom of this.  There are two issues:

 - first, you should more or less always give `--init` to containers.  Otherwise, your process ends up running as pid 1, with strange implications for signal reception.  Even a simple `podman run ... sleep 100` won't die with `^C` without `--init`.
 - (more complicated): the signal forwarding code in podman works by delivering any received signal to the `podman` command to the main command inside of the container (in our case, the shell).  In case of a long-running subcommand, `SIGINT` gets either blocked (busybox `ash`) or ignored (`bash`) by the shell, which means that pressing `^C` effectively doesn't work.  I found a neat trick to work around this: `npm install & wait`: that waits until `npm` is done, but does it in a context where the shell is able to receive the forwarded `SIGINT`.  In the event that the signal gets delivered, the shell will terminate, causing the runtime to take the entire container with it.